libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs

```diff
@@ -1229,623 +1230,1733 @@ private static (object, string) ParseISOTimestamp(string timeStamp, Func<DateTim
+                    {
+                        var convertedTZ = TimeZoneConverter.WindowsToIana(timezone);
+                        tz = TimeZoneInfo.FindSystemTimeZoneById(convertedTZ);
+                    }
+                    else
+                    {
+                        var convertedTZ = TimeZoneConverter.IanaToWindows(timezone);
+                        tz = TimeZoneInfo.FindSystemTimeZoneById(convertedTZ);
+                    }
+                }
+                catch
+                {
+                    error = "illegal timezone info";
                 }
 
-                if (!hasResult)
+                if (error == null)
                 {
-                    switch (num % 10)
+                    var convertedTS = TimeZoneInfo.ConvertTimeFromUtc(utcDt, tz);
+                    try
                     {
-                        case 1:
-                            ordinalResult += "st";
-                            break;
-                        case 2:
-                            ordinalResult += "nd";
-                            break;
-                        case 3:
-                            ordinalResult += "rd";
-                            break;
-                        default:
-                            ordinalResult += "th";
-                            break;
+                        result = convertedTS.ToString(format);
+                    }
+                    catch
+                    {
+                        error = "illegal format representation";
                     }
                 }
             }
 
-            return ordinalResult;
+            return (result, error);
         }
 
-        // conversion functions
-        private static string ToBinary(string strToConvert)
+        private static (string, string) ConvertToUTC(string sourceTimestamp, string sourceTimezone, string format = "o")
         {
-            var result = string.Empty;
-            foreach (var element in strToConvert.ToCharArray())
+            string error = null;
+            string result = null;
+            TimeZoneInfo tz = null;
+            var srcDt = DateTime.UtcNow;
+            try
             {
-                result += Convert.ToString(element, 2).PadLeft(8, '0');
+                srcDt = DateTime.Parse(sourceTimestamp);
+            }
+            catch
+            {
+                error = "illegal timestamp format";
             }
 
-            return result;
+            if (error == null)
+            {
+                try
+                {
+                    if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+                    {
+                        var convertedTZ = TimeZoneConverter.WindowsToIana(sourceTimezone);
+                        tz = TimeZoneInfo.FindSystemTimeZoneById(convertedTZ);
```

Comment:
`tz = TimeZoneInfo.FindSystemTimeZoneById(convertedTZ);` can be extracted out of if...else...